### ISSUE TITLE
Updates for GUI `v0.6.x`

### ISSUE THREAD

**Issue by jsiegle** [open-ephys-plugins/falcon-output#1]
Hi @marinechap -- the code in the [`api8-updates`](https://github.com/open-ephys-plugins/falcon-output/tree/api8-updates) branch is compatible with the latest version of the GUI. @anjaldoshi added the ability to sub-select channels from a given stream, so the whole buffer doesn't need to be sent at once. Can you test it out when you get a chance?

Also, the [`falcon-input`](https://github.com/open-ephys-plugins/falcon-output/tree/falcon-input) branch includes a new source plugin that can receive data from the Falcon Output, based on the C++ client example you wrote. However, I'm having trouble figuring out the right way to read data from separate channels. The code in [this loop](https://github.com/open-ephys-plugins/falcon-output/blob/96a196ab3c8f899de2fb86429f4b84164102d35d/Source/FalconInput.cpp#L226-L258) comes close, but it's still returning some zeros at the start of each buffer for channels > 0. Do you have any example code for unpacking data for multiple channels in C++?

**Comment by MarineChap** [comment]
Hello, I will test it today. Thanks for this development.

A few remarks already: I saw you changed the fbs schema. Adding field is fine and does not corrupt previous clients. But changing the name field (timestamps) means that all previous client needs to update their own fbs schema and client. 
If you think it is really important, no problem I can update on my side but it should be clear for all users that their client will break. 

I need to recheck but in first look your zero problem is not with the unpacking. The resizing step i was doing before 
```audioBuffer.setSize(nChannels, nSamples, true, true, true);``` disappeared. So I think you are never removing the zero before sending your packet.

I will do some tests and come back to you.

**Comment by MarineChap** [comment]
Feel like there is a big slowdown with this new version but not sure.

I would be happy to have your opinion on my new setup for latency computing. 

I am running falcon on the window computer with a docker image side by side with open ephys
in open ephys, 
I put a file reader / network events / falcon output / record node
in falcon - open ephys reader / ttl streamer / record node

Then I compare the timestamps in each falcon save packet (which is the num sample original from oe) with the sample number of the event arrival time in open-ephys. 
In the previous oe version, I had a 6ms delay at the start which kept along. In this version, the delay is very variant/huge and increase along the time. 

I put you the falcon graph / oe config in attached if you want to take a look. 
[oe_config.txt](https://github.com/open-ephys-plugins/falcon-output/files/9902831/oe_config.txt)
[OE_test_stream.txt](https://github.com/open-ephys-plugins/falcon-output/files/9902837/OE_test_stream.txt)

docker run -v c:/Users/neuropixels/Desktop/Marine/falcon:/root/log  -p 3335:3335 -p 5993:5993 -p 5555:5555 -p 5556:5556 -p 7777:7777 -it  marinechaput/falcon:test-oe

**Comment by jsiegle** [comment]
The reason we switched the naming conventions is because GUI `v0.6.x` makes a distinction between sample numbers (unsigned integers) and timestamps (floating-point values in seconds). We'd like to use the same conventions for this plugin, but we can switch it back if necessary.

We didn't think the call to `audioBuffer.setSize` was necessary, because the number of samples is passed to `openephysflatbuffer::CreateContinuousData`. Can you try adding it back to see if it makes a difference?

Also, you should double check to make sure the additional latency/jitter isn't due to compiling in Debug mode.

**Comment by MarineChap** [comment]
Could we stream both? I think that could be interesting to have one or the other available in the format and it won't add much data to the packet.
By not doing the audioBuffer.setSize, you are keeping the zeros in your buffer + giving the wrong size to the Create method meaning the packing/unpacking step will most likely corrupt the data. So no this step is super important. 
I will recheck the data today to verify it arrives correctly on the other side and keeps you posted.

For the latency, no, everything was compiled in release mode.  

(sorry for the delay, it is the holidays in Belgium/France these days)

**Comment by jsiegle** [comment]
I made some progress on this. I added back the `timestamp` field to the flatbuffer schema on the [`falcon-input`](https://github.com/open-ephys-plugins/falcon-output/tree/falcon-input) branch, which I'm using to send the software time at which the packet is sent. I can then query the software time on receive, which is always between 2 and 4 ms later for 384 channels of Neuropixels data read from a file. This is on the same machine, but at least it's consistent. Unpacking the buffer takes a bit longer, which explains the delay in the screenshot below.

![falcon-screenshot](https://user-images.githubusercontent.com/200366/200048875-c35dbd3b-0fd2-4cba-a2af-5220fa0c34a0.PNG)

In order to get all the samples to appear correctly on the other end, I had to create an intermediate buffer with the samples laid out consecutively. Passing the output of [AudioBuffer::getArrayOfReadPointers()](https://docs.juce.com/master/classAudioBuffer.html#a9cf2897fe524acc82e95337a033718a3) only works if the samples are contiguous in memory, which is not guaranteed. Using an intermediate buffer makes it possible to stream a random subset of channels, and doesn't seem to add much overhead.

Let me know when you're able to test things out with the new code!

**Comment by MarineChap** [comment]
Sound good. On what buffer size did you obtained this 2ms ? I tried to reproduce it with falcon on my computer and I see now I can go below the 3ms buffer. Do you know with npx if there is a minimal buffer size to use ? I guess too small would ask for too much cpu consumption. 

Also, I tried to work with timestamps but I am not sure to understand them. When streaming with the filereader, looks like it is starting at 49 sec, it is taking back the timestamps in the file not the real one right ? 
And what timestamps is streamed exactly ? What does this means ?   
``` double timestamp = double(Time::getHighResolutionTicks()) / double(Time::getHighResolutionTicksPerSecond());```
Thanks

**Comment by jsiegle** [comment]
We were using the default buffer size. In this case we're just measuring the time between packaging the ZMQ message and receiving it on the other end, so it's independent of the Open Ephys buffer. The minimal buffer size will depend on what your CPU can handle...I wouldn't try to push it below 5 ms.

The `timestamps` (in seconds) are based on the computer's hardware clock, which has sub-ms precision. However, these are only helpful if you're streaming data within the same machine. For inter-computer measurements, you could switch the `timestamps` to `Time::getCurrentTime().currentTimeMillis()`, assuming the system clocks are sync'd to within 1 ms.

The `sample_numbers` count the number of samples that have been read since the start of acquisition, rather than the original samples in the file.

You may have noticed that the latest version of the `falcon-input` branch includes an `event_codes` field that sends the state of the first 16 TTL channels. We've been able to use this to send TTL event states across multiple instances of the GUI.

**Comment by MarineChap** [comment]
Nice for the event_codes ! 

Can we have timestamps a bit more precise? Because now, with the standard buffer I have multiple times the same timestamps for the same buffer.  

```
68146.8
68146.8
68146.8
68146.9
68146.9
68146.9
68146.9
68147
68147
68147
68147
68147
68147.1
```
It is not usable info on falcon side. 

---

5ms buffer for applications like ripple disruption is a lot. Why would you not push below that? Because of the neuropixels hardware or open ephys and the computer using it? 

Have a good day

**Comment by MarineChap** [comment]
I think there is a bug with setting a different port. When it automatically (or from a file) reloads a chain with falcon output and a different port, I don't think this port is really set in background.

**Comment by jsiegle** [comment]
I'm not sure why the floats are getting rounded on the Falcon side, they are showing up with high precision inside the Falcon Input module in Open Ephys.

Regarding buffer size, a smaller buffer may be fine, you'd just have to make sure the CPU can keep up. Since one cycle of a 200 Hz ripple oscillation is ~5 ms, is it actually helpful to use a smaller buffer?

I will look into the port setting bug when I get back from SfN.

**Comment by MarineChap** [comment]
The value that I displayed was printed in open-ephys/falcon-output plugin. So falcon is not concerned by it.  I will look at it more closely, maybe just a rounding due to the std::cout. 

Enjoy SfN. Fabian went too.

**Comment by MarineChap** [comment]
Hello, Side note, you are still streaming zeros. I can understand it is easier like this for reading in falcon input but for any other clients, it is a problem to receive a packet with a certain size and then guess there are zeros to remove based on the nsamples value. We really need a resize to remove it before sending the packet.
Thanks.

**Comment by jsiegle** [comment]
I'm confused about how it could still be sending zeros. The latest code on the `falcon-input` branch creates an entirely new buffer of size `channels` x `samples`, so any trailing zeros in the `AudioBuffer` should be ignored. I'm wary of resizing the original buffer because it may be needed for other streams. Are you seeing a consistent number of zeros in each buffer, or does it vary?

Also, I'm trying to build the plugin on a new computer and CMake is failing in `FetchContent_MakeAvailable` because it **"could not find git for clone of flatbuffers-populate."** Googling around for help has not been helpful so far. Have you encountered this before?

**Comment by MarineChap** [comment]
Sorry for the side note ! You are right, I forget this rebufferization in the falcon input branch. Indeed no zeros ! thanks

About your bug, you don't have git on the computer. fetch content relies on git to clone and then install the library.

**Comment by jsiegle** [comment]
Ah, good call. Installing git fixed it!

The latest changes on the `falcon-input` branch ensure that the saved port number is set correctly.

Let us know if you have any other concerns, otherwise we will put out a new release.

**Comment by MarineChap** [comment]
Hello, when do you plan to merge this and release it ? I have some people waiting for it. 

For info, I won't be available for developing/testing before some time as I am taking some family time off. (also I am not working for the kloostermanLab anymore). 
Thank you

**Comment by jsiegle** [comment]
The new version will be ready soon, definitely before the end of Jan! We are just wrapping up the documentation.

**Comment by jsiegle** [comment]
This plugin is now available via the Plugin Installer in GUI v0.6.x! The updated documentation is here: https://open-ephys.github.io/gui-docs/User-Manual/Plugins/Falcon-Output.html

**Comment by MarineChap** [comment]
Amazing. Thank you  !
